polar-core/src/kb.rs

```diff
@@ -699,7 +699,20 @@ impl KnowledgeBase {
             }
         }
 
-        errors.into_iter().map(Diagnostic::Error).collect()
+        errors
+            .into_iter()
+            // Attach context to ResourceBlock errors.
+            //
+            // TODO(gj): can we attach context to *all* errors here since all errors will be
+            // parse-time errors and so will have some source context to attach?
+            .map(|e| match e.kind {
+                ErrorKind::Validation(ref e @ ValidationError::ResourceBlock { ref term, .. }) => {
```

Comment:
I think this is the only context-setting you need to keep. And yeah, I think this is potentially a reasonable place to pull in any additional context from the knowledge base.

Reply:
This only sets context for errors from `check_all_relation_types_have_been_registered()` and `as_rule()`.